File: Programmers/Lv3/travel_route.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<vector<string>> all_routes;

void dfs(vector<vector<string>> tickets, vector<bool> &visited,int cnt,int start, vector<string> &route){
    visited[start] = true;
    route.push_back(tickets[start][0]);
    
    /*모든 티켓을 사용했다면 */
    if(cnt == tickets.size()){
        route.push_back(tickets[start][1]);
        all_routes.push_back(route);
        route.pop_back();
        return ;
    }
    for(int i=0; i < tickets.size(); i++){
        /*방문하지 않았고 현재 항공권의 도착지와 다음 항공권의 출발지가 같다면*/
        if(!visited[i] && tickets[start][1] == tickets[i][0]){
            dfs(tickets,visited,cnt+1,i,route);
            visited[i] = false;
            route.pop_back();
        }
    }
}

vector<string> solution(vector<vector<string>> tickets) {
    vector<string> route;
    vector<bool> visited;
    int cnt=1;
    
    for(int i=0; i < tickets.size(); i++){
        if(tickets[i][0] == "ICN"){
            visited.assign(tickets.size(),false);
            route.clear();
            dfs(tickets,visited,cnt,i,route);
        }
    }
    sort(all_routes.begin(),all_routes.end());
    return all_routes[0];
}

File: java/src/programmers/lv3/TravelRoute.java
package programmers.lv3;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class TravelRoute {
    boolean[] visited;
    List<String> answer = new ArrayList<>();
    int N;
    public String[] solution(String[][] tickets) {
        N = tickets.length;
        visited = new boolean[N];

        for(int i=0; i < N; i++){
            if(tickets[i][0].equals("ICN")){
                visited[i] = true;
                dfs(tickets[i][1],2, tickets,"ICN");
                visited[i] = false;
            }
        }

        answer.sort(Comparator.comparing(String::toString));

        return answer.get(0).split(" ");
    }
    private void dfs(String city,int cnt, String[][] tickets,String route){
        route += " " + city;
        if(cnt == N+1){
            answer.add(route);
            return;
        }
        for(int i=0; i < N; i++){
            if(!visited[i] && tickets[i][0].equals(city)){
                visited[i] = true;
                dfs(tickets[i][1],cnt+1, tickets , route);
                visited[i] = false;
            }
        }
    }

    @Test
    void test(){
        Assertions.assertArrayEquals(new String[]{"ICN", "ATL", "ICN", "SFO", "ATL", "SFO"},
                solution(new String[][]{{"ICN", "SFO"}, {"ICN", "ATL"}, {"SFO", "ATL"}, {"ATL", "ICN"}, {"ATL","SFO"}}));
    }

    @Test
    void test1(){
        Assertions.assertArrayEquals(new String[]{"ICN", "JFK", "HND", "IAD"},
                solution(new String[][]{{"ICN", "JFK"}, {"HND", "IAD"}, {"JFK", "HND"}}));
    }
}
